src/Artifacts.jl

```diff
@@ -0,0 +1,594 @@
+        f()
+    finally
+        ARTIFACT_DIR_OVERRIDE[] = nothing
+    end
+end
+
+function artifacts_dir(args...)
+    if ARTIFACT_DIR_OVERRIDE[] === nothing
+        return abspath(depots1(), "artifacts", args...)
+    else
+        return abspath(ARTIFACT_DIR_OVERRIDE[], args...)
+    end
+end
+
+"""
+    create_artifact(f::Function)
+
+Creates a new artifact by running `f(artifact_path)`, hashing the result, and moving it
+to the artifact store (`~/.julia/artifacts` on a typical installation).  Returns the
+identifying tree hash of this artifact.
+"""
+function create_artifact(f::Function)
+    # Ensure the `artifacts` directory exists.
+    mkpath(artifacts_dir())
+
+    # Temporary directory where we'll do our creation business
+    temp_dir = mktempdir(artifacts_dir(); prefix="create_")
+
+    try
+        # allow the user to do their work inside the temporary directory
+        f(temp_dir)
+
+        # Calculate the tree hash for this temporary directory
+        artifact_hash = SHA1(tree_hash(temp_dir))
+
+        # If we created a dupe, just let the temp directory get destroyed. It's got the
+        # same contents as whatever already exists after all, so it doesn't matter.  Only
+        # move its contents if it actually contains new contents.
+        if !artifact_exists(artifact_hash)
+            # Move this generated directory to its final destination, set it to read-only
+            mv(temp_dir, artifact_path(artifact_hash))
+            set_readonly(artifact_path(artifact_hash))
+        end
+
+        # Give the people what they want
+        return artifact_hash
+    finally
+        # Always attempt to cleanup
+        rm(temp_dir; recursive=true, force=true)
+    end
+end
+
+"""
+    artifact_path(hash::SHA1)
+
+Given an artifact (identified by SHA1 git tree hash), return its installation path.
+"""
+artifact_path(hash::SHA1) = artifacts_dir(bytes2hex(hash.bytes))
+
+"""
+    artifact_exists(hash::SHA1)
+
+Returns whether or not the given artifact (identified by its sha1 git tree hash) exists
+on-disk.  If it does not, 
+"""
+artifact_exists(hash::SHA1) = isdir(artifact_path(hash))
+
+"""
+    remove_artifact(hash::SHA1)
+
+Removes the given artifact (identified by its SHA1 git tree hash) from disk.
+"""
+remove_artifact(hash::SHA1) = rm(artifact_path(hash); recursive=true, force=true)
+
+"""
+    verify_artifact(hash::SHA1)
+
+Verifies that the given artifact (identified by its SHA1 git tree hash) is installed on-
+disk, and retains its integrity. 
+"""
+function verify_artifact(hash::SHA1)
+    if !artifact_exists(hash)
+        return false
+    end
+
+    return hash.bytes == tree_hash(artifact_path(hash))
+end
+
+"""
+    archive_artifact(hash::SHA1, tarball_path::String)
+
+Archive an artifact into a tarball stored at `tarball_path`, returns the SHA256
+of the resultant tarball.  Throws an error if the artifact does not exist.
+"""
+function archive_artifact(hash::SHA1, tarball_path::String)
```

Comment:
```suggestion
function archive_artifact(hash::SHA1, destination_tarball_path::String)
```
or `save_path`?

Also this feels like it should be the first argument to me.

and also that the function should be `create_tarball(path, artifact::SHA1)` maybe?

Reply:
I like `archive_artifact()` better.  Since most of these artifact functions have the hash as the first argument, I think it's better to keep the consistency, despite the feeling that we are "modifying" the destination path.